### ISSUE TITLE
AES-GCM Exception Encrypting Large Files

### ISSUE THREAD

**Issue by vel0city** [pyca/cryptography#1821]
Using AES-GCM, I encounter an exception instead of an error message for the 2^39 - 256 byte limit.

    File "aes-gcm-test.py", line 25, in main
        encryptor.update(plaintext)
    File "...\hazmat\primitives\ciphers\base.py", line 159, in update
        return self._ctx.update(data)
    File "...\hazmat\backends\openssl\ciphers.py", line 118, in update
        assert res != 0
    AssertionError

I've been able to make this exception happen with the following code:

```
import os
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

def main():
    # Use AES-GCM
    encryptor = Cipher(algorithms.AES(os.urandom(16)), modes.GCM(os.urandom(12)), backend=default_backend()).encryptor()

    # 10MB of data
    plaintext = os.urandom(10000000)
    i = 0
    try:
        while True:
            encryptor.update(plaintext)
            i += 1
    except:
        print('{0}0MB encrypted before exception.'.format(i))
        raise

if __name__ == '__main__':
    main()
```

**Comment by alex** [comment]
a) If you're on OS X remember to make sure to use the OpenSSL backend to test thsi
b) It takes quite a while to encrypt enough data to hit the limit, not sure what the best testing strategy is.

**Comment by reaperhulk** [comment]
OpenSSL unfortunately doesn't give anything other than a 0 response code (no error on the stack). Looks like we're going to need to track the bytes encrypted. We already do this in parts of the CC backend so not too difficult. I'll put up a PR tomorrow for this.

**Comment by reaperhulk** [comment]
Anybody have an opinion what the exception should be?

**Comment by alex** [comment]
Value error seems on

**Comment by sbuss** [comment]
> Using AES-GCM, I encounter an exception instead of an error message for the 2^39 - 256 byte limit.

Note that according to the [NIST GCM spec](http://csrc.nist.gov/groups/ST/toolkit/BCM/documents/proposedmodes/gcm/gcm-spec.pdf) this is actually `2^39 - 256` *bits*, not bytes:

> A plaintext P, which can have any number of bits between 0 and 2^39 − 256. 

Also the spec limits AAD to 2^64 bits, so we need to track that as well:

> Additional authenticated data (AAD), which is denoted as A. This data is authenticated, but
not encrypted, and can have any number of bits between 0 and 2^64.

The plaintext and AAD limits will likely be separate PRs.

**Comment by alex** [comment]
Fixed by #2093.
